core/src/main/scala/org/apache/spark/scheduler/ResultTask.scala

```diff
@@ -66,7 +67,10 @@ private[spark] class ResultTask[T, U](
     _executorDeserializeTime = System.currentTimeMillis() - deserializeStartTime
 
     metrics = Some(context.taskMetrics)
-    func(context, rdd.iterator(partition, context))
+    val itr = rdd.iterator(partition, context)
+    val result = func(context, itr)
+    val computedFullPartition = itr.isEmpty || rdd.storageLevel != StorageLevel.NONE
```

Comment:
hmmm thats a good point - I'll move detecting computing the full partition being consumed to be per-RDD.